include/dali/core/mm/detail/free_list.h

```diff
@@ -0,0 +1,327 @@
+      if (gap_hi) {
+        // there's space at the end of the block - we need a new block to describe that
+        block *hi = get_block();
+        hi->next = lo->next;
+        lo->next = hi;
+
+        hi->start = aligned + bytes;
+        hi->end = lo->end;
+      }
+      // the space at the beginning of the block is the difference in alignments
+      lo->end = aligned;
+    } else {
+      if (gap_hi) {
+        // there's space at the end of the block - update current block to describe it
+        (*pbest)->start = aligned + bytes;
+      } else {
+        // the block was fully utilized - remove it from the free list
+        remove(pbest);
+      }
+    }
+    return aligned;
+  }
+
+  void put(void *ptr, size_t bytes) {
+    block *blk = get_block();
+    blk->start = static_cast<char*>(ptr);
+    blk->end = blk->start + bytes;
+    blk->next = head_;
+    head_ = blk;
+  }
+
+ protected:
+  /**
+   * @brief Recycle an unused block descriptor or create a new one.
+   */
+  block *get_block() {
+    if (unused_blocks_) {
+      block *blk = unused_blocks_;
+      unused_blocks_ = blk->next;
+      blk->next = nullptr;
+      return blk;
+    } else {
+      return new block();
+    }
+  }
+
+  /**
+   * @brief Remove a block from the list and place it in unused_blocks_
+   *
+   * @param ppblock pointer to a block pointer - this automatically updates head_
+   */
+  void remove(block **ppblock) {
+    block *b = *ppblock;
+    block *next = b->next;
+    *ppblock = next;
+    b->start = nullptr;
+    b->end = nullptr;
+    b->next = unused_blocks_;
+    unused_blocks_ = b;
+  }
+
+  block *head_ = nullptr;
+  block *unused_blocks_ = nullptr;
+};
+
+
+/**
+ * @brief Maintains a list of free memory blocks of variable size, returning free blocks
+ *        with least margin.
+ *
+ * This free list tries to connect a free block to its immediate predecessor and successor.
+ * This causes the list to represent true fragmentation - all contiguous free blocks are stored
+ * as such.
+ *
+ * There is an extra cost of going through the list upon placing a memory region in the list,
+ * but this should be offset by the list being typically shorter than a non-coalescing one.
+ *
+ * Free time is increased compared to non-coalescing variant, but allocation time is reduced
+ * due to reduced element count.
+ */
+class coalescing_free_list : public best_fit_free_list {
+ public:
+  coalescing_free_list() = default;
+  coalescing_free_list(const coalescing_free_list &) = delete;
+  coalescing_free_list(coalescing_free_list &&other) {
+    swap(other);
+  }
+
+  coalescing_free_list &operator=(coalescing_free_list &&other) {
+    static_cast<best_fit_free_list&>(*this) = std::move(other);
+    return *this;
+  }
+
+  void put(void *ptr, size_t bytes) {
+    // find blocks that immediatele precede and succeed the freed block
+    block *pred = nullptr;
+    block **succ = nullptr;
+    for (block **pb = &head_; *pb; pb = &(*pb)->next) {
```

Comment:
Does it make any sense to add assert to check if there are two block that are reported as pred or succ?

Reply:
It does, but it will be partial - the loop stops when if find both.

Reply:
Done.